### PR TITLE
Add Ext4 support and serve ext2 mounts through it

### PR DESCRIPTION
**Pull request by byliangg**
## Summary

This PR grows the ext2 driver into an Ext4 implementation that keeps
serving ext2 volumes, the way Linux does (`CONFIG_EXT4_USE_FOR_EXT2`): the ext4
driver gains an ext2-compatible indirect block-mapping engine, takes over
the `ext2` filesystem-type name, and the classic ext2 module is retired
(about 8.5k lines removed; the PR is net -5k).

The end state: one driver serves both `mount -t ext2` and `-t ext4`;
`mkfs.ext2` volumes keep their on-disk format bit-for-bit (verified with
e2fsck and Linux cross-reads); the full ext regression suite and xfstests
pass under both type names.

## Review guide (4 commits)

1. **Rename the ext2 module to ext4** — mechanical only (directory and
   identifiers; the registered type name stays `ext2`, zero behavior
   change). Done first so every later change reads as a file-level diff
   against the existing driver.
2. **Rework the driver core around a unified block-mapping architecture** —
   inode metadata via slot-level RMW honoring the volume's `inode_size`
   (serialized per group: slots are sub-sector), a per-inode
   `BlockMapping` dispatch seam (indirect engine only at this point),
   engine-agnostic direct IO, and carried-over xattr/special files/
   statfs/mount options. A `MountFlavor` records the mount name; only
   `ext2` is registered yet.
3. **Add the extent block-mapping engine and register the ext4 type** —
   the extent tree engine plugs into the seam, each inode selects its
   engine by its own `EXTENTS` flag (Linux's rule), and the `ext4` name
   is registered with `IS_EXT2_SB`-style gating. The kernel test suite
   (192 tests) lands here with dual-format fixtures.
4. **regression/ext: run one suite across a mount-flavor matrix** — the
   per-driver test directories unify into `regression/fs/ext/` (git
   tracks them as renames), executed over a feature matrix of `mke2fs`
   images under every legal type name, plus negative mount-rejection
   cells.

## Test-suite unification

Following the suggestion in the earlier review comment, the per-driver
test directories are unified into `regression/fs/ext/`: one set of C
tests, one `ext.tests` list, executed against a feature matrix of images
(`mke2fs -t ext2` in three shapes, ext4 with/without journal/extents) and
under both mount flavors, plus negative cells asserting mount rejection
(no filetype, rev0, metadata_csum, and journal-under-ext2).

## Verification

- Kernel ktests: 191 passing (mount matrix, slot I/O, engines, xattr,
  fallocate, statfs, symlink/device edge cases).
- Guest regression: the shared ext suite over every matrix image x every
  legal mount flavor; all green.
- xfstests (`FSTYP=ext2`, full runlist, unchanged configuration):
  **Passed all 82 tests**, identical case-for-case to the pre-change
  baseline captured on the same machine. An additional run mounting the
  same ext2-format disks under the `ext4` name also passes all 82.
- `e2fsck -fn` clean on every image after the runs; Linux reads back the
  volumes written by this driver.
- Every commit builds and passes `make check` individually.

## Behavior changes vs. the retired ext2 driver (all fail-closed)

- Volumes with a journal, rev0/no-filetype volumes: previously accepted
  by the ext2 driver (and could be misparsed); now rejected under
  `-t ext2` (journaled images mount as `-t ext4`, journal ignored).
- Unknown ro_compat features (e.g. `metadata_csum`): previously silently
  ignored on a read-write mount; now rejected.
- sync no longer rewrites backup superblocks/GDTs (matches Linux runtime
  behavior; backups belong to mkfs/fsck/resize).
- Inode-slot writes honor the on-disk `inode_size` (128-byte-inode
  volumes no longer risk clobbering the neighboring slot).

## Stated limitations

- No journaling/JBD2: clean journaled volumes mount with the journal
  ignored and are written without it; `RECOVER` (unreplayed journal) is
  rejected. Crash consistency is future work.
- 4 KiB filesystem blocks only; no dir_index (htree), no metadata_csum,
  no 64bit/flex_bg/inline_data (all rejected at mount, not silently
  ignored).
- `last_orphan` processing is a no-op seam (orphan space is reclaimed by
  fsck, as before).

### CODE REVIEW

**Comment by cchanging** [comment]
We currently have a dedicated test directory for ext2, but many of those tests should also be applicable to ext4. Is there a way to let ext4 reuse some of the existing ext2 tests? For example, we could refactor things along the following lines:

```text
fs/ext/
├── Makefile
├── fs_test.h
├── ext2.tests
├── ext4.tests
├── file_io.c
├── namei.c
├── rename.c
└── ...
```

In `fs_test.h`, define a filesystem-neutral test root:

```c
#ifndef FS_TEST_H
#define FS_TEST_H

#define FS_TEST_ROOT "/ext-test"

#endif
```

Then update the test sources from:

```c
#define BASE_DIR "/ext2/file_io_test"
```

to:

```c
#include "fs_test.h"

#define BASE_DIR FS_TEST_ROOT "/file_io_test"
```

After that, we could update `regression/fs/run_tesh.sh` and adjust the `test_ext2`-related logic so that the mounted ext2/ext4 directory is first bind-mounted to `ext-test`, and then the corresponding tests are executed.

We could also maintain separate allowlists for each filesystem:

```text
# ext2.tests
fallocate
file_io
mknod
namei
open_dir
open_unlink
permissions
readdir
rename
rmdir
short_rw
sparse
symlink
unix_socket
xattr
```

```text
# ext4.tests
file_io
namei
open_dir
open_unlink
permissions
readdir
rename
rmdir
sparse
symlink
```

One principle here is that we should try to maximize reasonable code and test reuse, while avoiding filesystem-specific handling whenever possible. You could also think about whether there is a better way to refactor and share these tests, or whether the naming could be improved. In the future, this approach could even be extended to support tests for exfat as well.

File: kernel/src/fs/fs_impls/ext4/mod.rs [hunk 1]
```diff
@@ -1,72 +1,26 @@
...
-//! Directory entry layout helpers live under `inode::dir::dir_entry`,
-//! next to the directory operations that use them.
-//!
-//! # References
-//!
-//! - <https://www.kernel.org/doc/html/latest/filesystems/ext2.html>
-//! - <https://www.nongnu.org/ext2-doc/ext2.html>
+//! The implementation supports extent-based and ext2-style indirect file I/O,
+//! linear directories, special files, and extended attributes. Journaling,
+//! metadata checksums, and indexed directories are not supported yet.
```
Comment (cchanging) [inline kernel/src/fs/fs_impls/ext4/mod.rs:7]: A module’s documentation should describe what functionality the module implements, briefly introduce that functionality, and possibly explain the semantics of some key structs. It should not describe your work plan or unfinished parts. You can refer to the module-level documentation in `kernel/comps/virtio/src/device/socket/mod.rs` as an example.

File: kernel/src/fs/fs_impls/ext4/mod.rs [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+mod inode;
+mod prelude;
+mod super_block;
+mod utils;
+mod xattr;
+
+#[cfg(ktest)]
+mod test_utils;
+
+/// Registers the ext4 and ext2 filesystem types with the VFS registry; the
```
Comment (cchanging) [inline kernel/src/fs/fs_impls/ext4/mod.rs:36]: "the one driver serves both names." is unnecessary.

File: kernel/src/fs/fs_impls/ext4/super_block.rs [hunk 3]
```diff
@@ -1,166 +1,75 @@
...
+use super::{
+    block_group::RawBlockGroup,
+    feature::{
+        FeatureCompatSet, FeatureIncompatSet, FeatureRoCompatSet, INCOMPAT_SUPP, RO_COMPAT_SUPP,
+    },
+    prelude::*,
+    utils::IsPowerOf,
+};
+
+/// Magic signature (`s_magic`).
```
Comment (cchanging) [inline kernel/src/fs/fs_impls/ext4/super_block.rs:17]: What is `s_magic`? In the document, do not mention the detail of Linux implementation, and don't use the concept that has not been explained.

File: kernel/src/fs/fs_impls/ext4/super_block.rs [hunk 4]
```diff
@@ -1,166 +1,75 @@
...
+/// The main superblock is located at byte 1024 from the start of the device.
 pub(super) const SUPER_BLOCK_OFFSET: usize = 1024;
 
 const SUPER_BLOCK_SIZE: usize = 1024;
 
-/// Validated, Rust-typed in-memory representation of the ext2 superblock.
+/// Validated, Rust-typed in-memory representation of the ext4 superblock.
+///
+/// Block counts use the implementation's 64-bit block-number type, although
+/// the `64BIT` on-disk feature is not supported yet.
```
Comment (cchanging) [inline kernel/src/fs/fs_impls/ext4/super_block.rs:28]: Again, do not describe what you have not done in the document.

File: kernel/src/fs/fs_impls/ext4/super_block.rs [hunk 5]
```diff
@@ -1,166 +1,75 @@
...
     free_inodes_count: u32,
-    /// First data block.
+    default_reserved_uid: u32,
+    default_reserved_gid: u32,
     first_data_block: Ext4Bid,
-    /// Block size.
     block_size: usize,
-    /// Fragment size.
-    frag_size: usize,
-    /// Number of blocks in each block group.
```
Comment (cchanging) [inline kernel/src/fs/fs_impls/ext4/super_block.rs:66]: Why remove the documents for each field?

File: kernel/src/fs/fs_impls/ext4/super_block.rs [hunk 6]
```diff
@@ -1,166 +1,75 @@
...
         }
         if sb.log_frag_size != sb.log_block_size {
             return_errno_with_message!(Errno::EINVAL, "invalid fragment size");
         }
-
         let block_size = BLOCK_SIZE;
-        let frag_size = BLOCK_SIZE;
 
-        let state = FsState::from_bits(sb.state)
-            .ok_or(Error::with_message(Errno::EINVAL, "invalid fs state"))?;
```
Comment (cchanging) [inline kernel/src/fs/fs_impls/ext4/super_block.rs:163]: I am not sure whether this error return will disappear in ext4. Have you test it?

File: kernel/src/fs/fs_impls/ext4/super_block.rs [hunk 7]
```diff
@@ -171,42 +80,60 @@ impl TryFrom<RawSuperBlock> for SuperBlock {
...
-                let inode_size = sb.inode_size as usize;
-                if inode_size < 128 {
-                    return_errno_with_message!(Errno::EINVAL, "inode size is too small");
-                }
-                if inode_size > BLOCK_SIZE {
-                    return_errno_with_message!(Errno::EINVAL, "inode size is too large");
-                }
-                if !inode_size.is_power_of_two() {
-                    return_errno_with_message!(Errno::EINVAL, "inode size is not power of two");
+                let inode_size = usize::from(sb.inode_size);
```
Comment (cchanging) [inline kernel/src/fs/fs_impls/ext4/super_block.rs:91]: Why change `as usize` to `usize::from`?

File: kernel/src/fs/fs_impls/ext4/super_block.rs [hunk 8]
```diff
@@ -171,42 +80,60 @@ impl TryFrom<RawSuperBlock> for SuperBlock {
...
-            return_errno_with_message!(Errno::EINVAL, "inodes per group is too small");
-        }
-
-        let max_bits_per_group = (block_size as u32) * 8;
-        if nr_inodes_per_group > max_bits_per_group {
-            return_errno_with_message!(Errno::EINVAL, "inodes per group is too large");
+        let inodes_per_block =
+            u32::try_from(block_size / inode_size).expect("filesystem block size fits u32");
+        let max_bits_per_group =
+            u32::try_from(block_size).expect("filesystem block size fits u32") * 8;
```
Comment (cchanging) [inline kernel/src/fs/fs_impls/ext4/super_block.rs:134]: Why use `try_from`? This transform will not fail.

File: kernel/src/fs/fs_impls/ext4/feature.rs [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+        /// `RO_COMPAT_METADATA_CSUM`: crc32c checksums on metadata.
+        const METADATA_CSUM = 1 << 10;
+        /// `RO_COMPAT_PROJECT`.
+        const PROJECT = 1 << 13;
+        /// `RO_COMPAT_VERITY`.
+        const VERITY = 1 << 15;
+    }
+}
+
+/// Incompatible features this implementation supports.
```
Comment (cchanging) [inline kernel/src/fs/fs_impls/ext4/feature.rs:104]: Do not say "this implementation". After being merged into the main branch, readers who see “this implementation” will not know what implementation it refers to. In this case, it is better to simply state what is currently supported.

File: kernel/src/fs/fs_impls/ext4/super_block.rs [hunk 10]
```diff
@@ -240,303 +167,157 @@ impl TryFrom<RawSuperBlock> for SuperBlock {
...
 
-    /// Returns the fragment size.
+    /// Returns the fragment size. Mount validation requires it to equal the
+    /// block size, so no separate field is kept.
     pub(super) const fn fragment_size(&self) -> usize {
-        self.frag_size
+        self.block_size
     }
 
-    /// Returns total number of inodes.
```
Comment (cchanging) [inline kernel/src/fs/fs_impls/ext4/super_block.rs:382]: Avoid making deletions or changes that are not self-explanatory.

File: kernel/src/fs/fs_impls/ext4/fs.rs [hunk 11]
```diff
@@ -596,443 +695,570 @@ impl Ext4 {
...
-        raw_sb: &RawSuperBlock,
-        sb_offset: usize,
-        group_desc_bid: Ext4Bid,
-    ) -> Result<()> {
-        let group_desc_segment = self.group_descriptors_segment.clone();
-        let bio_segment = BioSegment::new_from_segment(group_desc_segment, BioDirection::ToDevice);
-        self.write_blocks(group_desc_bid, bio_segment)
-            .map_err(|_| {
-                Error::with_message(Errno::EIO, "failed to write group descriptor table")
+        let mut sb = self.super_block.write();
```
Comment (cchanging) [inline kernel/src/fs/fs_impls/ext4/fs.rs:704]: Is this change correct?

Originally, the `Ext2` `SuperBlock` abstraction maintained all fields of `RawSuperBlock`, so during `sync` it could be directly converted into a `RawSuperBlock` and written back. After this change, however, each `sync` first has to read from disk, then modify only a small subset of fields, and write it back. Intuitively, this feels like incorrect behavior.

**Review by cchanging — COMMENTED** [review]
Thank you for your effort. At the moment, this PR still has quite a few issues.

One aspect is the documentation: you can take in some of the points from the comments below and then go through the documentation as a whole again.

Another issue is that this PR contains too many changes to the existing code that are hard to justify, such as unexplained deletions, unexplained replacements with equivalent code, and even some changes that seem to make the code worse rather than better. I think the actual changes needed for this PR should be much smaller than what is currently presented.

As the PR author, you should try to be accountable for every change you make, or at least understand the reason behind each one. Since this PR is relatively large, I understand that this is not easy, but I still hope you can do your best to achieve that.
